Added `build.build-dir` support

ranger-ross opened the pull request:
This PR adds `build_dir` to the `BuildConfig`

See: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#build-dir

closes #28

taiki-e commented:
Thanks! The implementation looks good.

However, given that this is an unstable feature and could lead to errors if the accepted syntax changes in the future, it might be better to put it behind a feature flag.

ranger-ross commented:
Sure, perhaps we add a generic `unstable` flag similar to [cargo_metadata](https://github.com/oli-obk/cargo_metadata/blob/main/Cargo.toml#L25)

taiki-e commented:
That sounds like a good idea.

I'm usually against using the Cargo feature for unstable stuff (https://github.com/rust-lang/api-guidelines/pull/284#issuecomment-2809562060), but in this case I think the Cargo feature will be fine since the only users affected are those who explicitly use the unstable feature in their config files.

ranger-ross commented:
I added an `unstable` feature flag and support for it in `tools/codegen`.
let me know what you think :)

ranger-ross commented:
Added `CARGO_BUILD_BUILD_DIR` handling in the latest push.

Regarding the failing CI jobs, seems unrelated.
```
Preparing a careful sysroot (target: x86_64-unknown-linux-gnu)... thread 'main' panicked at src/main.rs:267:10:
failed to build sysroot; run `cargo careful setup` to see what went wrong: sysroot build failed
```

I pulled the latest nightly and run `cargo +nightly careful test --all --all-features` on my machine and could not reproduce a failure.

taiki-e commented on src/easy.rs:
```diff
@@ -564,6 +571,8 @@ impl BuildConfig {
         });
         let target_dir = de.target_dir.map(|v| v.resolve_as_path(current_dir).into_owned());
+        #[cfg(feature = "unstable")]
+        let build_dir = de.build_dir.map(|v| v.resolve_as_path(current_dir).into_owned());
```
Hmm... given the presence of templates, I wonder if there is a problem with unconditional invocation of `resolve_as_path` here.

Cargo Reference says:


> This option supports path templating.
> 
> Available template variables:
> 
> - `{workspace-root}` resolves to root of the current workspace.
> - `{cargo-cache-home}` resolves to `CARGO_HOME`
> - `{workspace-path-hash}` resolves to a hash of the manifest path

Perhaps `resolve_as_path` should be skipped if the path starts with `{workspace-root}` or `{cargo-cache-home}`.

Also, it would be nice to mention in the documentation that these templates are not resolved.

ranger-ross replied:
yes, agreed. 
Added handling for that in the latest commit.

ranger-ross commented:
Sorry for the late follow up on this!

Addressed your comment and pushed another commit to stabilize `build-dir`.
I left the unstable feature machinery for future use, though I can remove it you'd prefer.

taiki-e reviewed (approved):
Thanks!
